storage_controller/src/service.rs

```diff
@@ -1637,21 +1626,36 @@ impl Service {
                     | NodeSchedulingPolicy::Filling
             );
 
-            if !node.is_available() || reset_scheduling {
-                let mut new_nodes = (**nodes).clone();
-                if let Some(node) = new_nodes.get_mut(&reattach_req.node_id) {
+            let mut new_nodes = (**nodes).clone();
+            if let Some(node) = new_nodes.get_mut(&reattach_req.node_id) {
+                if reset_scheduling {
+                    node.set_scheduling(NodeSchedulingPolicy::Active);
+                }
+
+                if response.tenants.len() <= 5 {
```

Comment:
Lets' come up with a less arbitrary way to make the initial set of pageservers go active immediately: we could go directly to Active for the first two pageservers added, perhaps?  i.e. we can't do much useful without two pageservers, so there's no point waiting around until we're adding "extra" pageservers.

For tests that use more than two pageservers, we could make the test code wait for their statuses to reach active -- that way we avoid a test-specific hack in the controller code.

Reply:
I agree we need a better heuristic. Ideally it would consider the sum of the working set size in bytes over the tenants to reattach.

---

I'm not sure the node count is strictly better. It would impact any prod regions where the storcon has 2 pageservers under management. Assume these pageservers are loaded, and the drain did not manage to move tenants away from the ps being restarted (maybe they were pinned or didn't have secondaries). In this case we'd mark the node active and it would get the short HB timeout resulting in the node flaking.

---

> we could make the test code wait for their statuses to reach active -- that way we avoid a test-specific hack in the controller code.

Could just do that and remove the heuristic altogether?

Reply:
After much test massaging, the test suite passes without the heuristic.